Fix compiler warnings pertaining to uniform_int()

imaginary-person opened the pull request:
**PROBLEM DESCRIPTION:**
GitHub issue 46391 suggests that compiler warnings pertaining to _floating-point value does not fit in required integral type_ might cause some confusion.

These compiler-warnings arise during compilation of the templated function `uniform_int()`. The warnings are misleading because they arise from the way the compiler compiles templated functions, but the if-else statements in the function obviate the possibilities that the warnings describe. So, the purpose of a fix would only be to fix the compiler warnings, and not to fix any sort of a bug.


**FIX DESCRIPTION:**
[EDITED, after inputs from @malfet]: In the function `uniform_int()`, the if-else conditions pertaining to types `double` & `float` can be removed, and then an overloaded specialized function can be added for floating-point types. The current version of the function can be specialized to not have its return type as a floating point type.

An unrelated observation is that the if-else condition pertaining to the type `double` (line 57 in the original code) was redundant, as line 61 in the original code covered it (`std::is_floating_point<T>::value` would also have been true for the type `double`).

Fixes #46391

imaginary-person committed:
Fix compiler warnings pertaining to uniform_int()

PROBLEM DESCRIPTION:
GitHub issue 46391 suggests that compiler warnings pertaining to "floating-point value does not fit in required integral type" might cause some confusion.

These compiler-warnings arise during compilation of the templated function uniform_int(). The warnings are misleading because they arise from the way the compiler compiles templated functions, but the if-else statements in the function obviate the possibilities that the warnings describe. So, the purpose of a fix would only be to fix the compiler warnings, and not to fix any sort of a bug.


FIX DESCRIPTION:
In the function uniform_int(), the if-else conditions pertaining to types double & float can be removed,
and then uniform_int() can be overloaded for both double & float.
Since template specialization for functions cannot be partial, we would have to add four overloaded versions of uniform_int(typename T, typename V) pertaining to 
type T being either double or float, and type V being either uint32_t or uint64_t.

An unrelated observation is that the if-else condition pertaining to the type 'double' (line 57 in the original code) was redundant, as line 61 in the original code covered it (std::is_floating_point<T>::value would also have been true for the type 'double').

malfet reviewed (approved):
Thank you for the fix!

malfet commented on aten/src/ATen/core/TransformationHelper.h:
```diff
@@ -68,6 +66,46 @@ C10_HOST_DEVICE inline T uniform_int(V val) {
 }
 
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <float, uint32_t>
+ */
+template <>
+C10_HOST_DEVICE inline float uniform_int(uint32_t val) {
+  return static_cast<float>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<float>::digits) + 1));
+}
+
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <float, uint64_t>
+ */
+template <>
+C10_HOST_DEVICE inline float uniform_int(uint64_t val) {
+  return static_cast<float>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<float>::digits) + 1));
+}
+
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <double, uint32_t>
+ */
+template <>
+C10_HOST_DEVICE inline double uniform_int(uint32_t val) {
+  return static_cast<double>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<double>::digits) + 1));
+}
+
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <double, uint64_t>
```
Small nits:
 - Can you fold 4 comment blocks into 4?
 - This implies that those functions are only called for unsigned integer type. Is it possible to add a compile time check that those templates could not be instantiated with signed integral type?
 P.S. Also, please explain why something like the following wouldn't work (instead of 4 explicit full specializations):
```
template<typename T, typename V>
C10_HOST_DEVICE inline std::enable_if<std::is_floating_point<T>::value, T>::type uniform_int(V val) {
    return static_cast<T>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<T>::digits) + 1));
}

imaginary-person replied:
Thank you for your prompt review, and for this learning opportunity!

I had never used enable_if before, so I agree that your suggestion is way better than using full specializations.

As for the compile-time check for disallowing instantiation with a signed integral type, could you please suggest what should the behavior be in that case? Please advise if an assertion would be appropriate. Thanks!

imaginary-person committed:
Fix compiler warnings related to uniform_int()

Based on @malfet's advice, used std::enable_if instead of using four full specializations

imaginary-person committed:
Removed extraneous empty line

I had introduced an empty line by mistake. Removed it

imaginary-person committed:
Fix compiler warnings related to uniform_int()

As per malfet's advice, overloaded uniform_int() for floating-point types & non-floating-point types to prevent compiler warnings reported via GitHub issue 46391.

Didn't change any code from the last commit in the forked repo, but only edited a comment, so as to re-run tests automatically on Circle CI, and check if any reported build failures on Circle CI were due to this commit, although it seems highly unlikely.

imaginary-person commented:
> ##  CI failures summary and remediations
> As of commit [64c7d6d](https://github.com/pytorch/pytorch/commit/64c7d6d0f686371993ee0c7aa57e15da02ab2037) (more details [on the Dr. CI page](https://dr.pytorch.org/commit-details.html?sha1=64c7d6d0f686371993ee0c7aa57e15da02ab2037)):
> 
> * **1/1** failures possibly[*](https://code.facebook.com/ci/dr-ci-info/) introduced in this PR
>   
>   * **1/1** non-CircleCI failure(s)
> 
> ### ci.pytorch.org: 1 failed
> * **Failed:** [`pr/pytorch-linux-bionic-rocm3.10-py3.6`](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-trigger/1210/)
> 
> This comment was automatically generated by [Dr. CI](https://code.facebook.com/ci/dr-ci-info/) (expand for details).

pr/pytorch-linux-bionic-rocm3.10-py3.6 is actually failing due to another reason, as the failure-cause is same in some of the previous builds that didn't have this change.

imaginary-person reviewed (commented):
Hello @malfet, based on your advice, I had revised the code to have one overloaded version of `uniform_int(T, V) `that corresponds to `T` being a floating point type. As for the original function, I ensured that `T` won't be a floating point type.

The sole failed check reported by Jenkins yesterday was unrelated to the change in this PR & seems to have been fixed now. Can you please check if this fix is acceptable? Thanks!

Snippets from the commit -
```
/**
 * A transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`.
 * In order to prevent compiler warnings reported in GitHub issue 46391, T can't be float or double
 * in this overloaded version
 */
template <typename T, typename V>
C10_HOST_DEVICE inline typename std::enable_if<!(std::is_floating_point<T>::value), T>::type uniform_int(V val) {
```

```
/**
 * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
 * added to fix compiler warnings reported in GitHub issue 46391. T is either float or double in this version.
 */
template<typename T, typename V>
C10_HOST_DEVICE inline typename std::enable_if<std::is_floating_point<T>::value, T>::type uniform_int(V val) {
  return static_cast<T>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<T>::digits) + 1));
}
```
